atrium-repo/src/mst.rs

```diff
@@ -0,0 +1,1347 @@
+        ///
+        /// We deserialize this with the [`Ipld`] type instead of directly as a `Vec<u8>`,
+        /// because serde maps the latter to CBOR Major Type 4 (array of data items) instead
+        /// of Major Type 2 (byte string). Other crates exist that provide bytes-specific
+        /// deserializers, but `Ipld` is already in our dependencies.
+        #[serde(rename = "k")]
+        pub key_suffix: Ipld,
+
+        /// ("value", CID Link, required): link to the record data (CBOR) for this entry.
+        #[serde(rename = "v")]
+        pub value: Cid,
+
+        /// ("tree", CID Link, nullable): link to a sub-tree [`Node`] at a lower level which
+        /// has keys sorting after this [`TreeEntry`]'s key (to the "right"), but before the
+        /// next [`TreeEntry`]'s key in this [`Node`] (if any).
+        #[serde(rename = "t")]
+        pub tree: Option<Cid>,
+    }
+}
+
+/// Merkle search tree helper algorithms.
+mod algos {
+    use super::*;
+
+    pub enum TraverseAction<R, M> {
+        /// Continue traversal into the specified `Cid`.
+        Continue((Cid, M)),
+        /// Stop traversal and return `R`.
+        Stop(R),
+    }
+
+    /// Compute the depth of the specified node.
+    ///
+    /// If both the node and its nested subtrees do not contain leaves, this will return `None`.
+    pub async fn compute_depth(
+        mut bs: impl AsyncBlockStoreRead,
+        node: Cid,
+    ) -> Result<Option<usize>, Error> {
+        // Recursively iterate through the tree until we encounter a leaf node, and then
+        // use that to calculate the depth of the entire tree.
+        let mut subtrees = vec![(node, 0usize)];
+
+        loop {
+            if let Some((subtree, depth)) = subtrees.pop() {
+                let node = Node::read_from(&mut bs, subtree).await?;
+                if let Some(layer) = node.layer() {
+                    return Ok(Some(depth + layer));
+                }
+
+                subtrees.extend(node.trees().cloned().zip(std::iter::repeat(depth + 1)));
+            } else {
+                return Ok(None);
+            }
+        }
+    }
+
+    /// Traverse a merkle search tree.
+    ///
+    /// This executes the closure provided in `f` and takes the action
+    /// returned by the closure.
+    /// This also keeps track of "seen" nodes, and if a node is seen twice, traversal
+    /// is immediately halted and an error is returned.
+    pub async fn traverse<R, M>(
+        mut bs: impl AsyncBlockStoreRead,
+        root: Cid,
+        mut f: impl FnMut(Node, Cid) -> Result<TraverseAction<R, M>, Error>,
+    ) -> Result<(Vec<(Node, M)>, R), Error> {
+        let mut node_cid = root;
+        let mut node_path = vec![];
+        let mut seen = HashSet::new();
+
+        loop {
+            let node = Node::read_from(&mut bs, node_cid).await?;
+            if !seen.insert(node_cid) {
+                // This CID was already seen. There is a cycle in the graph.
```

Comment:
Is it even possible to serialize a graph with a cycle?
The two nodes would point at each other, and by definition it should be impossible to compute their CIDs if there is a cycle.

Reply:
So I actually don't think this is possible, which means the extra validation logic here is probably optional and very unlikely to be hit in practice.
To err on the side of safety though, I think it would be best to leave this in-place.